### ISSUE TITLE
WString pointers cause issues on stack for certain API calls

### ISSUE THREAD

**Issue by jcspencer** [rodrigocfd/winsafe#97]
Hi,

In testing `WString`, I've come across an issue where if the input is <20 characters and the underlying buffer is stored on the stack, certain Win32 APIs fail to correctly read the string contents.

An example of where this occurs is in [HEVENTLOG](https://github.com/rodrigocfd/winsafe/blob/master/src/kernel/handles/heventlog.rs#L65); if the contents of the string is <20 characters, the underlying buffer is allocated on the stack, and the event log output becomes corrupted.

I'm not sure the best way to solve this, but perhaps adding a flag to force usage of the heap in buffer allocations could work?

Thanks for all your work on this library!

**Comment by rodrigocfd** [comment]
Hi,

Did you find another place where this problem occurs, or it's just this specific `HEVENTLOG` call?

**Comment by jcspencer** [comment]
Not so far; I've only really been working with HEVENTLOG.

From testing, it seems padding a string with trailing NUL bytes up to 20 to skip the fast-path on the stack works around the problem (which points to an issue with stack pointers), but I haven't had a chance to test other functions.

**Comment by rodrigocfd** [comment]
I found a minor bug in the multi-string conversion, where the null separators were not being properly inserted, apart from the double terminating null.

Please let me know if this solved your issue.

**Comment by jcspencer** [comment]
Hi @rodrigocfd, thanks for the quick response!

On some further digging, it seems that [`HEVENTLOG::ReportEvent`](https://github.com/rodrigocfd/winsafe/blob/master/src/kernel/handles/heventlog.rs#L46) actually takes an pointer of an _array of WString pointers_, rather than a buffer containing chained WStrings with double null termination.

This isn't the nicest way to do it, but for example:

```rust
// Turn a Option<Vec<String>> into Vec<WString>
let strings: Vec<WString> = match arguments {
    Some(arg) => {
        arg
        .iter()
        .map(|s| WString::from_str(s))
        .collect()
    },
    None => Vec::new()
};

// Construct a Vec of pointers to the WString arguments.
let msg_array: Vec<*const u16> =
    strings
    .iter()
    .map(|s| s.as_ptr())
    .collect();

bool_to_sysresult(
    unsafe {
        ReportEventW(
            self.handle.ptr() as isize,
            severity.raw(),
            0,
            event_id,
            std::ptr::null_mut(),
            msg_array.len() as _,
            0,
            msg_array.as_ptr() as _,
            std::ptr::null_mut(),
        )
    },
)
```

**Comment by rodrigocfd** [comment]
Yes, you're correct. I've implemented [`CreateProcess`](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw) recently, and it takes a block of null-terminated strings with a terminating double-null. I though it was the same case, but it's indeed a different raw format.

As a side note, `Vec::as_ptr` for an empty `Vec` [doesn't return null](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1437a1f14183eddbcf5ef79bfe898e38), so we must handle the empty case explicitly.

I also reviewed all other cases where an array of strings is needed, and `ReportEvent` seems to be the only one with this peculiarity. So let me know if it's working now.

Thank you very much.
